### PR TITLE
Use Bundler::Audit::Scanner for analysis

### PR DESCRIPTION
**Pull request by dblandin**
This PR updates the engine analysis to use bundler-audit's internal `Scanner` class which offers better access to gem and advisory information.

This particular change is also necessary to disable the gem's network access attempt in order to determine whether a Gemfile source is internal or external, an update which will be introduced following this PR.

This PR also takes a few refactoring steps to improve overall structure and clarify remediation point calculation.

There should be no behavioral changes as a result of the updates in this PR.

@codeclimate/review :mag_right:

**Before**

```json
{
    "categories": [
        "Security"
    ],
    "check_name": "Insecure Dependency",
    "content": {
        "body": "**Advisory**: CVE-2013-1800\n\n**Criticality**: High\n\n**URL**: http://osvdb.org/show/osvdb/90742\n\n**Solution**: upgrade to >= 0.3.2"
    },
    "description": "crack Gem for Ruby Type Casting Parameter Parsing Remote Code Execution",
    "location": {
        "lines": {
            "begin": 87,
            "end": 87
        },
        "path": "Gemfile.lock"
    },
    "remediation_points": 500000,
    "severity": "critical",
    "type": "Issue"
}

```

**After**

```json
{
    "categories": [
        "Security"
    ],
    "check_name": "Insecure Dependency",
    "content": {
        "body": "**Advisory**: CVE-2013-1800\n\n**Criticality**: High\n\n**URL**: http://osvdb.org/show/osvdb/90742\n\n**Solution**: upgrade to >= 0.3.2"
    },
    "description": "crack Gem for Ruby Type Casting Parameter Parsing Remote Code Execution",
    "location": {
        "lines": {
            "begin": 87,
            "end": 87
        },
        "path": "Gemfile.lock"
    },
    "remediation_points": 500000,
    "severity": "critical",
    "type": "Issue"
}
```

### CODE REVIEW

File: lib/cc/engine/bundler_audit/analyzer.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          @directory = directory
+          @io = io
+        end
+
+        def run
+          if gemfile_lock_exists?
+            Dir.chdir(directory) do
+              Bundler::Audit::Scanner.new.scan do |vulnerability|
+                result = Result.new(vulnerability, File.open(gemfile_lock_path))
+                issue = result.to_issue
```
Comment (GordonDiggs) [inline lib/cc/engine/bundler_audit/analyzer.rb]: WDYT of changing `to_issue` below to `as_json` and then you could just do `result.to_json` below

File: lib/cc/engine/bundler_audit/result.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        def initialize(result, gemfile_lock)
+          @gem = result.gem
+          @advisory = result.advisory
+          @gemfile_lock = gemfile_lock
+        end
+
+        def to_issue
+          {
+            categories: ["Security"],
```
Comment (GordonDiggs) [inline lib/cc/engine/bundler_audit/result.rb]: :lipstick: use `%w`

File: lib/cc/engine/bundler_audit/analyzer.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        def initialize(directory:, io: STDOUT)
+          @directory = directory
+          @io = io
+        end
+
+        def run
+          if gemfile_lock_exists?
+            Dir.chdir(directory) do
+              Bundler::Audit::Scanner.new.scan do |vulnerability|
+                result = Result.new(vulnerability, File.open(gemfile_lock_path))
```
Comment (GordonDiggs) [inline lib/cc/engine/bundler_audit/analyzer.rb]: Seems less than efficient to do `File.open` every time. Worth it to pull out into a method?

File: lib/cc/engine/bundler_audit/result.rb (hunk 2 above)
Comment (pbrisbin) [inline lib/cc/engine/bundler_audit/result.rb]: FWIW: I think our rubocop config uses the default which recommends `%w` only at a certain size (2, 3?). If we intend to use it at any size we should update that config.

File: lib/cc/engine/bundler_audit/result.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
+module CC
+  module Engine
+    module BundlerAudit
+      class Result
```
Comment (GordonDiggs) [inline lib/cc/engine/bundler_audit/result.rb]: I would call this class `Issue`

Comment (GordonDiggs) [inline lib/cc/engine/bundler_audit/result.rb]: This class could use some tests, too

File: lib/cc/engine/bundler_audit/remediation.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
+module CC
+  module Engine
+    module BundlerAudit
+      class Remediation
```
Comment (GordonDiggs) [inline lib/cc/engine/bundler_audit/remediation.rb]: This class could use some tests

File: lib/cc/engine/bundler_audit/analyzer.rb (hunk 1 above)
Comment (dblandin) [inline lib/cc/engine/bundler_audit/analyzer.rb]: I think I'd feel more comfortable with that if we renamed the class to `Issue`. Then we would have:

```ruby
issue = Issue.new(vulnerability, File.open(gemfile_lock_path))
io.print("#{issue.to_json}\0")
```

WDYT?

**Comment by GordonDiggs** [comment]
I like this direction. Did we lose the insecure source checks entirely?

File: lib/cc/engine/bundler_audit/analyzer.rb (hunk 3 above)
Comment (dblandin) [inline lib/cc/engine/bundler_audit/analyzer.rb]: Ahh, good point. This should only be opened once.

File: lib/cc/engine/bundler_audit/result.rb (hunk 4 above)
Comment (dblandin) [inline lib/cc/engine/bundler_audit/result.rb]: I'm personally okay with the higher level test including an actual `Gemfile.lock` file on Analyzer here.

File: lib/cc/engine/bundler_audit/remediation.rb (hunk 5 above)
Comment (dblandin) [inline lib/cc/engine/bundler_audit/remediation.rb]: Definitely.

File: lib/cc/engine/bundler_audit/result.rb (hunk 2 above)
Comment (dblandin) [inline lib/cc/engine/bundler_audit/result.rb]: I was planning to update the rubocop config with our styleguide in a separate PR and address any CC issues that come up as a result. Sound good or should I do that in this PR?

**Comment by dblandin** [comment]
> I like this direction. Did we lose the insecure source checks entirely?

I don't think those checks ever worked as the output looks completely different and isn't picked up by our currently handling of the output on master.

Here's an example of that output. We currently just ignore it. I'd like to bring in proper support for insecure source checks in a separate PR.

```
"Insecure Source URI found: http://rubygems.org/\nVulnerabilities found!\n"
```

**Comment by GordonDiggs** [comment]
>  I'd like to bring in proper support for insecure source checks in a separate PR.

Sounds good to me!

**Comment by dblandin** [comment]
@codeclimate/review This is ready for another :mag_right:

File: .codeclimate.yml [hunk 6]
```diff
@@ -1,6 +1,11 @@
...
   rubocop:
     enabled: true
+    exclude_fingerprints:
+    # Ignoring long method length for Issue#to_json
+    - 3d618821b1ce28599d6c54f90bb4df59
 ratings:
   paths:
   - "**.rb"
+exclude_paths:
+  - bin/bundler_audit.rb
```
Comment (GordonDiggs) [inline .codeclimate.yml]: Why are we excluding this file?

Comment (dblandin) [inline .codeclimate.yml]: Rubocop generated an error on the shebang line. I think we usually drop the extension on executables in other engines, which is why this doesn't often need to be excluded. Might be worth doing that instead.

Comment (dblandin) [inline .codeclimate.yml]: Never mind. That doesn't seem to be the issue. I get the false positive locally but not on dotcom. Strange.

```
❯ codeclimate analyze -e rubocop                                         codeclimate-bundler-audit/git/devon/use-bundler-audit-scanner-for-analysis
Starting analysis
Running rubocop: Done!

== bin/bundler-audit (1 issue) ==
1: Use snake_case for source file names. [rubocop]

Analysis complete! Found 1 issue.
```

**Comment by dblandin** [comment]
@GordonDiggs Updated to remove that exclude path.

**Comment by GordonDiggs** [comment]
LGTM!

**Pull request merged**